dev/src/cubegen.cpp

```diff
@@ -559,134 +559,160 @@ nfr("cg_load_vox", "name", "S", "R:voxels]S?",
+                        if (ReadSpanVec<string, int32_t>(p, key))
+                            if (key == "_name") {
+                                string value;
+                                if (ReadSpanVec<string, int32_t>(p, value))
+                                    node_names.insert_or_assign(node_id, value);
+                            } else
+                                SkipSpanVec<string, int32_t>(p);
                     }
-                    auto models_num = ReadMemInc<int32_t>(c);
-                    for (int i = 0; i<models_num; ++i) {
-                        auto model_id = ReadMemInc<int32_t>(c);
-                        node_to_model.insert_or_assign(node_id, model_id);
-                        auto dict_len = ReadMemInc<int32_t>(c);
-                        for (int i = 0; i<dict_len; ++i) {
-                            SkipVec<string, int32_t>(c);
-                            SkipVec<string, int32_t>(c);
+                    int32_t models_num;
+                    if (ReadSpanInc(p, models_num))
+                        for (int i = 0; i < models_num; ++i) {
+                            int32_t model_id;
+                            if (ReadSpanInc(p, model_id))
+                                node_to_model.insert_or_assign(node_id, model_id);
+                            int32_t dict_len;
+                            if (ReadSpanInc(p, dict_len))
+                                for (int i = 0; i < dict_len; ++i) {
+                                    SkipSpanVec<string, int32_t>(p);
+                                    SkipSpanVec<string, int32_t>(p);
+                                }
+                            [[maybe_unused]] int32_t reserved;
+                            ReadSpanInc(p, reserved);
                         }
-                        [[maybe_unused]]auto reserved = ReadMemInc<int32_t>(c);
-                    }
                 } else if (!strncmp(id, "LAYR", 4)) {
                     // Layer metadata
-                    auto c = (const uint8_t*)p;
-                    auto layer_id = ReadMemInc<int32_t>(c);
-                    auto dict_len = ReadMemInc<int32_t>(c);
-                    for (int i = 0; i<dict_len; ++i) {
+                    int32_t layer_id;
+                    if (!ReadSpanInc(p, layer_id)) return erreof();
+                    int32_t dict_len;
+                    if (!ReadSpanInc(p, dict_len)) return erreof();
+                    for (int i = 0; i < dict_len; ++i) {
                         string key;
-                        ReadVec<string, int32_t>(c, key);
-                        if (key == "_name") {
-                            string value;
-                            ReadVec<string, int32_t>(c, value);
-                            layer_names.insert_or_assign(layer_id, value);
-                        } else
-                            SkipVec<string, int32_t>(c);
+                        if (ReadSpanVec<string, int32_t>(p, key))
```

Comment:
here you're simply skipping data if eof, whereas above its an error, is that intentional?

Reply:
i was a bit worried about the cost in a loop, but its a branch anyways so that doesn't make sense in the current code, changed.